File: manager/src/main/java/moe/shizuku/manager/ShizukuApplication.java
package moe.shizuku.manager;

import android.app.Application;
import android.content.Context;
import android.os.Build;

import com.topjohnwu.superuser.Shell;

import me.weishu.reflection.Reflection;
import moe.shizuku.manager.authorization.AuthorizationManager;
import rikka.material.app.DayNightDelegate;
import rikka.material.app.LocaleDelegate;

public class ShizukuApplication extends Application {

    static {
        Shell.Config.setFlags(Shell.FLAG_REDIRECT_STDERR);
    }

    public static Context getDeviceProtectedStorageContext(Context context) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            return context.createDeviceProtectedStorageContext();
        }
        return context;
    }

    public static void init(Context context) {
        ShizukuSettings.initialize(context);
        LocaleDelegate.setDefaultLocale(ShizukuSettings.getLocale());
        DayNightDelegate.setApplicationContext(context);
        DayNightDelegate.setDefaultNightMode(ShizukuSettings.getNightMode());
        AuthorizationManager.init(context);
    }

    @Override
    public void onCreate() {
        super.onCreate();

        init(this);
    }

    @Override
    protected void attachBaseContext(Context base) {
        super.attachBaseContext(base);
        Reflection.unseal(base);
    }
}

File: manager/src/main/java/moe/shizuku/manager/ShizukuApplication.kt
package moe.shizuku.manager

import android.app.Application
import android.content.Context
import android.os.Build
import androidx.appcompat.app.AppCompatDelegate
import com.topjohnwu.superuser.Shell
import moe.shizuku.manager.ktx.logd
import org.lsposed.hiddenapibypass.HiddenApiBypass
import rikka.core.util.BuildUtils.atLeast30
import rikka.material.app.LocaleDelegate

lateinit var application: ShizukuApplication

class ShizukuApplication : Application() {

    companion object {

        init {
            logd("ShizukuApplication", "init")

            Shell.setDefaultBuilder(Shell.Builder.create().setFlags(Shell.FLAG_REDIRECT_STDERR))
            if (Build.VERSION.SDK_INT >= 28) {
                HiddenApiBypass.setHiddenApiExemptions("")
            }
            if (atLeast30) {
                System.loadLibrary("adb")
            }
        }
    }

    private fun init(context: Context?) {
        ShizukuSettings.initialize(context)
        LocaleDelegate.defaultLocale = ShizukuSettings.getLocale()
        AppCompatDelegate.setDefaultNightMode(ShizukuSettings.getNightMode())
    }

    override fun onCreate() {
        super.onCreate()
        application = this
        init(this)
    }

}
